nomad/job_endpoint.go

```diff
@@ -1908,6 +1909,40 @@ func (j *Job) Dispatch(args *structs.JobDispatchRequest, reply *structs.JobDispa
 		dispatchJob.Meta[k] = v
 	}
 
+	// Ensure that we have only one dispatched version of this job running concurrently
+	// by comparing the idempotency token against any non-terminal versions.
+	if args.IdempotencyToken != "" {
+		// Fetch all jobs that match the parameterized job ID prefix
+		iter, err := snap.JobsByIDPrefix(ws, parameterizedJob.Namespace, parameterizedJob.ID)
+		if err != nil {
+			return fmt.Errorf("failed to retrieve jobs for idempotency check")
+		}
+
+		// Iterate
+		for {
+			raw := iter.Next()
+			if raw == nil {
+				break
+			}
+
+			// Ensure the parent ID is an exact match
+			existingJob := raw.(*structs.Job)
+			if existingJob.ParentID != dispatchJob.ParentID {
+				continue
+			}
+
+			// Idempotency tokens match. Ensure existing job is terminal.
+			if existingJob.DispatchIdempotencyToken == args.IdempotencyToken {
+				// The existing job is either pending or running.
+				// Registering a new job would violate the idempotency token.
+				if existingJob.Status != structs.JobStatusDead {
```

Comment:
Great point, I hadn't considered the short dispatch case. For our use case on HCP, the idempotency token will map to an action for a specific resource (ie `consul-cluster/{uuid}/instance-create`). In the case of a transient error in the dispatched job, we'd like to retry immediately, using the same idempotency token. If idempotency is enforced for dead jobs waiting to be GC'd we wouldn't be able to retry as quickly as we'd prefer.

Reply:
Oh I see. So there are two types of transient failures:

First, the dispatch request fails with network transient error (e.g. network or timeout) where the dispatch request may have failed but may also have succeed but the caller wasn't informed of the success. The API consumer will want to retry with idempotent token to avoid repeating the side-effects. I assume here we want to enforce the token for the longest reasonable time, as it never hurts.

The second failure is if the dispatch request succeeded but the dispatched job failed with a transient failure (e.g. instance creation failed for transient error and should be retried). Here, api consumer want to retry as soon as the job failed. 

Is the second failure case that's driving this design? I would have hoped that the [`restart` block](https://www.nomadproject.io/docs/job-specification/restart) addressed this case without the API consumers retrying dispatching jobs manually.

Also, you can achieve the second failure handling semantics even if idempotent token is enforced for dead jobs. e.g. you can have `consul-cluster/{resource uuid}/instance-create/{request uuid}`, where `{request uuid}` is updated when the dispatch job failed and you want to retry creating the same resource.

Reply:
Ah yes, totally forgot about the `restart` block and your `{request uuid}` suggestion makes perfect sense so I will remove the check for job status and update the comment to say something about the job not being GC'd yet.